File: sample/product/cpplib/test/python/testDates.py
import unittest
import datetime

import py_clib as clib

class TestDates(unittest.TestCase):
    def setUp(self): pass
    def tearDown(self): pass
    def testDayOfWeek(self):
        """Tests clib.DayOfWeek function"""
        date = datetime.date(2014,10,4) # which is Saturday
        print "DayOfWeek(%s) = %d" % (date, clib.DayOfWeek(date))
        for dow in range(7):
            date += datetime.timedelta(1)
            self.assertEqual(clib.DayOfWeek(date), dow)
            pass #end-for
        pass #end-def
    def testDaysInMonth(self):
        """Tests clib.DaysInMonth function"""
        normalYear = 2014
        leapYear = 2016
        for month in range(1,13):
            normalDays = clib.DaysInMonth(normalYear, month)
            leapDays   = clib.DaysInMonth(leapYear, month)

            if month == 2: self.assertEqual(leapDays - normalDays, 1)
            else: self.assertEqual(leapDays, normalDays)

            if month in [1,3,5,7,8,10,12]:
                self.assertEqual(normalDays, 31)
            elif month in [4,6,9,11]:
                self.assertEqual(normalDays, 30)
            elif month == 2:
                self.assertEqual(normalDays, 28)
            else: raise Exception("PROGRAM BUG")
            pass #end-for
        pass #end-def
    def testDateToEOM(self):
        """Tests clib.DateToEOM and related functions"""
        for year in range(2015,2017):
            for month in range(1,13):
                daysInMonth = clib.DaysInMonth(year, month)
                date        = datetime.date(year, month, 15)
                endOfMonth  = clib.DateToEOM(date)
                self.assertEqual(endOfMonth, datetime.date(year, month, daysInMonth))
                pass #end-for
            pass #end-for
        pass #end-def
    def testDateIntervalFreq(self):
        """Tests clib.DateIntervalFreq"""
        # DateInterval is a simple type and we convert freely to/from strings
        frequencies = [12, 6, 4, 3, 2, 1]
        for frequency in frequencies:
            months = 12 / frequency
            interval = "%dM" % months
            freq = clib.DateIntervalFreq(interval)
            print interval, freq, frequency
            self.assertEqual(freq, frequency)
            pass #end-for
        pass #end-def
    def testDateIntervalYears(self):
        """Tests clib.DateIntervalYears"""
        # DateInterval is a simple type and we convert freely to/from strings
        years = [0.5, 1, 1.5, 2, 3, 4, 5]
        months = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
        weeks = [1, 2, 3]
        days = [1, 2, 3, 7, 14]

        for year in years:
            if year == float(int(year)):
                interval = "%dY" % int(year)
            else: interval = "%dM" % int(12 * year)
            result = clib.DateIntervalYears(interval)
            print interval, result
            self.assertEqual(result, year)
            pass #end-for

        for month in months:
            interval = "%dM" % month
            result = clib.DateIntervalYears(interval)
            print interval, result
            self.assertEqual(result, month/12.0)
            pass #end-for

        for week in weeks:
            interval = "%dW" % week
            result = clib.DateIntervalYears(interval)
            print interval, result
            self.assertEqual(result, 7*week/365.0)
            pass #end-for

        for day in days:
            interval = "%dD" % day
            result = clib.DateIntervalYears(interval)
            print interval, result
            self.assertEqual(result, day/365.0)
            pass #end-for
        pass #end-def
    pass #end-class

if __name__ == "__main__":
    clib.start_logging("dates.log")
    unittest.main()
    pass #end-if
    

File: sample/product/clib/test/cpp/testDates.cpp
/*
***************************************************************************
** Tests of all the functions and classes defined in dates.cfg
***************************************************************************
*/

#include "clib_dateType.hpp"
#include "clib_dates.hpp"
#include "clib_clib_service.hpp"

#include <string>
#include <sstream>
#include <iostream>
#include <vector>
#include <stdio.h>

#include <spi/TestRunner.hpp>
#include <spi/RuntimeError.hpp>

static void testDayOfWeek(void)
{
    spi::Date date = clib::Date(2014,10,5); // which is Sunday
    std::cout << "DayOfWeek(" << date.ToString() << ") = " << clib::DayOfWeek(date) << std::endl;

    // Monday is day of week zero

    for (int dow = 0; dow < 7; ++dow)
    {
        date += 1;

        clib::Weekday weekday = clib::DayOfWeek(date);
        clib::Weekday::Enum eWeekday = weekday;

        std::cout << date.ToString() << " : " << weekday << " : " << eWeekday << " : " << dow << std::endl;

        SPI_POST_CONDITION(eWeekday == dow);
    }
}

static void testDaysInMonth(void)
{
    int normalYear = 2014;
    int leapYear = 2016;
    for (int month = 1; month <= 12; ++month)
    {
        int normalDays = clib::DaysInMonth(normalYear, month);
        int leapDays   = clib::DaysInMonth(leapYear, month);

        if (month == 2)
        {
            SPI_POST_CONDITION(leapDays - normalDays == 1);
        }
        else
        {
            SPI_POST_CONDITION(leapDays == normalDays);
        }

        switch (month)
        {
        case 1:
        case 3:
        case 5:
        case 7:
        case 8:
        case 10:
        case 12:
            SPI_POST_CONDITION(normalDays == 31);
            break;
        case 4:
        case 6:
        case 9:
        case 11:
            SPI_POST_CONDITION(normalDays == 30);
            break;
        case 2:
            SPI_POST_CONDITION(normalDays == 28);
            break;
        default:
            throw spi::RuntimeError("PROGRAM_BUG");
        }
    }
}

static void testDateToEOM(void)
{
    for (int year = 2015; year <= 2016; ++year)
    {
        for (int month = 1; month <= 12; ++month)
        {
            int daysInMonth = clib::DaysInMonth(year, month);
            spi::Date date = clib::Date(year, month, 15);
            spi::Date endOfMonth;

            clib::DateToEOM(date, endOfMonth);

            SPI_POST_CONDITION(endOfMonth = clib::Date(year, month, daysInMonth));
        }
    }
}

static void testDateIntervalFreq(void)
{
    // DateInterval is a simple type and we convert freely to/from strings
    int frequencies[] = {12, 6, 4, 3, 2, 1};

    for (size_t i = 0; i < 6; ++i)
    {
        int frequency = frequencies[i];
        int months = 12 / frequency;
        std::string interval = spi_util::StringFormat("%dM", months);
        // note that DateIntervalConstSP can be constructed implicitly from std::string
        double freq = clib::DateIntervalFreq(interval);

        std::cout << interval << " " << freq << " " << frequency << std::endl;

        SPI_POST_CONDITION(freq == frequency);
    }
}

static void testDateIntervalYears(void)
{
    // DateInterval is a simple type and we convert freely to/from strings
    double years[] = {0.5, 1, 1.5, 2, 3, 4, 5};
    int months[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12};
    int weeks[] = {1, 2, 3};
    int days[] = {1, 2, 3, 7, 14};

    for (size_t i = 0; i < sizeof(years)/sizeof(double); ++i)
    {
        char interval[16];
        double year = years[i];
        if (year == (double)(int)year)
        {
            sprintf(interval, "%dY", (int)year);
        }
        else
        {
            sprintf(interval, "%dM", (int)(year*12));
        }
        // note that DateIntervalConstSP can be constructed implicitly from std::string
        // and hence from const char[16] as well
        double result = clib::DateIntervalYears(interval);

        std::cout << interval << " " << result << std::endl;

        SPI_POST_CONDITION(result == year);
    }

    for (size_t i = 0; i < sizeof(months)/sizeof(int); ++i)
    {
        char interval[16];
        int month = months[i];
        sprintf(interval, "%dM", month);

        double result = clib::DateIntervalYears(interval);

        std::cout << interval << " " << result << std::endl;

        SPI_POST_CONDITION(result == month/12.0);
    }

    for (size_t i = 0; i < sizeof(weeks)/sizeof(int); ++i)
    {
        char interval[16];
        int week = weeks[i];
        sprintf(interval, "%dW", week);
        double result = clib::DateIntervalYears(interval);

        std::cout << interval << " " << result << std::endl;

        SPI_POST_CONDITION(result == 7*week/365.0);
    }

    for (size_t i = 0; i < sizeof(days)/sizeof(int); ++i)
    {
        char interval[16];
        int day = days[i];
        sprintf(interval, "%dD", day);

        double result = clib::DateIntervalYears(interval);

        std::cout << interval << " " << result << std::endl;

        SPI_POST_CONDITION(result == day / 365.0);
    }
}

int main(int argc, char* argv[])
{
    clib::clib_exported_service()->start_logging("dates.log", "");
    return spi::TestRunner(argc, argv,
                           "DayOfWeek", testDayOfWeek,
                           "DaysInMonth", testDaysInMonth,
                           "DateToEOM", testDateToEOM,
                           "DateIntervalFreq", testDateIntervalFreq,
                           "DateIntervalYears", testDateIntervalYears,
                           NULL);
}
